src/libgit2/config.c

```diff
@@ -593,36 +593,67 @@ static int get_backend_for_use(git_config_backend **out,
 	git_config *cfg, const char *name, backend_use use)
 {
 	size_t i;
+	size_t len;
 	backend_internal *backend;
+	int error = 0;
+	git_config_entry *entry = NULL;
+	char *key = NULL;
 
 	*out = NULL;
 
-	if (git_vector_length(&cfg->backends) == 0) {
+	len = git_vector_length(&cfg->backends);
+	if (len == 0) {
 		git_error_set(GIT_ERROR_CONFIG,
 			"cannot %s value for '%s' when no config backends exist",
 			uses[use], name);
 		return GIT_ENOTFOUND;
 	}
 
 	git_vector_foreach(&cfg->backends, i, backend) {
-		if (!backend->backend->readonly) {
-			*out = backend->backend;
-			return 0;
+		if (backend->backend->readonly)
+			continue;
+
+		/* git-config doesn't update worktree-level config
+		   unless specifically requested; follow suit. If you
+		   specifically want to update that level, open the
+		   single config level with git_config_open_level and
+		   provide that as the config. In this case, there
+		   will only be one backend in the config. */
+		if (len > 1 && backend->level == GIT_CONFIG_LEVEL_WORKTREE)
+			continue;
+
+		/* If we're trying to delete a piece of config, make
+		   sure the backend we return actually defines it in
+		   the first place. */
+		if (use == BACKEND_USE_DELETE) {
```

Comment:
I need to go through and refresh my understanding of our configuration logic. This change _makes sense_ if the perspective of libgit2 is _when we say delete a configuration key, then just delete it wherever it is_. (And I think that's the semantics that our current config data sort of implies.)

This is different than `git config`, which _takes_ a config entry to operate on.

But I think that this change actually suggests a deeper problem with our API. IIUC, the goal of this change is to go find the config that we should delete a key in. What should we do if there's a multivar split over two configuration files? Instead `get_backend_for_use`, should we iterate over all backends?

(Not asking you to make this change, asking for your opinion on the API.)

Reply:
> What should we do if there's a multivar split over two configuration files? 

Certainly this case would surprise folks; I would not expect to get different results with repeated calls:

```c
git_config_delete_entry(cfg, "foo.bar"); // deletes --local
git_config_delete_entry(cfg, "foo.bar"); // deletes --global
```

That just feels 'wrong'. It would feel much more appropriate to say

```c
git_config_delete_entry_at_level(cfg, GIT_CONFIG_LEVEL_LOCAL, "foo.bar"); // deletes --local
git_config_delete_entry_at_level(cfg, GIT_CONFIG_LEVEL_GLOBAL, "foo.bar"); // deletes --global
// with companion functions
git_config_set_string_at_level(cfg, GIT_CONFIG_LEVEL_LOCAL, "foo.bar", "baz");
// etc.
```

It was surprising to us to learn that the current config API did not work this way. In a vacuum, I would consider `git_config_set_string_at_level` to be the correct abstraction for a C library: explicit and unambiguous even if a little verbose.

I feel comfortable enough with the config code at this point to tackle such a thing in a separate PR if that's a direction you'd like to go. (In fact, I've rather been enjoying this codebase – it's a nice change of pace for me 😄.) It probably deserves a more robust design process, though.

It would also make sense to me to incorporate the worktree-config support *only* into that new API if that helps resolve concerns with this PR, but that will come with different considerations for doing due diligence to ensure consumer applications continue to 'work reasonably correctly'. (Although maybe that's not as high of a concern in this context as long as we make the appropriate deprecation warnings; I've not worked with this particular kind of project in a FOSS setting before.)

Reply:
I'm not positive but my thinking here is that maybe we should keep things mostly as they are but, yes, avoid surprise:

I think that the "simple" `git_config` functions should probably mirror the git command line itself. Reads return the "effective" configuration (applying the config files in order of priority and giving a result.) But we change writes/deleted to always refer to the local.

If somebody wants a specific level then they can open that file and operate on it directly. I don't think we necessarily need APIs to read and write at a specific level directly, I think that people can probably go get the config file that they're interested in.

But this rather presupposes that git's model is right here — and I'm increasingly convinced that it actually is.

The rather tricky part is the *user* configuration(s). We may need to be more thoughtful than "always write to local".

Reply:
I was reviewing this comment once more before writing up what this would look like

> I think that the "simple" git_config functions should probably mirror the git command line itself. Reads return the "effective" configuration (applying the config files in order of priority and giving a result.) But we change writes/deleted to always refer to the local.

I think that's a good instinct as it would follow some principle of least surprise.

What I'm not sure of is what is meant by this:

> The rather tricky part is the _user_ configuration(s). We may need to be more thoughtful than "always write to local".

By 'user configurations', do you mean things like `--global` and `--system` config?

Here's what I'm going to propose:

> For reads: keep the current implementation and use the most specific level available to us.
> 
> For writes:
> 1. If there is exactly one config backend, use it.
> 2. If `--local` is available, use it.
> 3. Signal an error (as `git` does).

The alternative to (3) (which might be the 'more thoughtful' approach, as best as I can imagine one) would be to set the most specific level of config available to us if `--local` is not available, but I rather think that might be too smart for this 'plumbing' function. I do think the need of 'set this at the system/global level' would be met with `git_config_open_level` followed by `git_config_set_string` following the logic proposed above. (Some wrapper 'porcelain' could be devised as above, but it does seem like that would be low-value API.)

Reply:
I have pushed c82bef33a090217e731944e91385e1f3095c1fee which takes the approach outlined above. Before the final version, I would like to add a test that `git_config_open_level(&out, cfg, GIT_CONFIG_LEVEL_LOCAL)` only ever returns a config with one backend (even in degenerate cases) since I'm now relying on that behavior in the base case of `get_backend_for_use`.

Note that tests introduced in #4217 are non-trivially altered (@pks-t), but I believe the test is still effective in demonstrating that the `readonly` flag is respected.

There is another problem `test_remote_httpproxy__env` that I've not tracked down yet; some details are in the commit though.

Reply:
I spent a little time staring at this over the weekend and there's much more complexity here than I expected. You may have identified some of these problems as well. In particular, backends are shared across multiple `git_config` instances. (The repository has a configuration with multiple backends with an ordering. When you create a snapshot, you get a new configuration with read only backends. And you can `open_level`, you get a new configuration object with only one of those backends.)

The other "gotcha" is that we support users of the library to create their own custom configuration, `GIT_CONFIG_LEVEL_APP` (or higher) which is higher-priority than the system/global/etc configuration. And we have no idea what they're doing. So trying to be clever about where writes go needs to take into account the `APP` level which may or may not exist and may or may not be writeable.

The current flow today is "write to the highest level backend that isn't readonly", which _may_ be `APP` and if it's not it _will_ be `LOCAL`. 

So, because of this shared state, you can't simply make the worktree backend "not writable" for the repository configuration, since it will also be not writable when you `open_level` it, which you _do_ want to write to. 🙃 

And trying to hardcode any logic about defaulting to the local configuration is tricky, since there might be `APP`-levels that you need to contend with (and that are being written to today and we would like to preserve backcompat).

My initial thought was that we could re-use the `readonly` bit (or possibly make a new "writes need to be explicit" bit) and make the worktree config readonly when it's created via `git_repository_open`, but that doesn't work, because subsequent `open_level` will happily just give you the existing worktree config which is... readonly.

Instead, I'm coming back to the notion that configuration needs separate read and write ordering. Providing a `set_writeorder` API allows callers (in particular, `git_repository`) to set that they want `LOCAL` to be the only write target during configuration creation. Users can subsequently add an `APP` level, which will be be the highest-level writer (but they can change that themselves with `set_writeorder`).

Ultimately, `set_writeorder` only applies to the things that are in your config object. So if you want to override the default write order of the thing(s) you just added, you will need to call it again after adding them. (You can't reconfigure your write order.)

This required us to pull apart the refcounted backend object from the list of backends that are in a config so that they can be separate. Alas. In any case, I pushed up a branch [`ethomson/worktree-config`](https://github.com/libgit2/libgit2/tree/ethomson/worktree-config) ([diff](https://github.com/libgit2/libgit2/compare/main...ethomson/worktree-config)). I'd love your feedback.

I also found and fixed some interesting bugs, notably that you can't `git_transaction_free` an uncommitted transaction today. And that if you start a transaction and then mess around with backends, you're going to have a bad time.

Reply:
I think that we were typing our comments at roughly the same time. 😅 

Right, so, my concern here is definitely around the `CONFIG_LEVEL_APP` no longer being a write target. I think that's critical to backcompat... Our CLI uses it to map `-c` command-line arguments into a configuration backend (though this is read-only). I know that at least one hosting provider uses `CONFIG_LEVEL_APP`, and ... I have no idea if they're using it to write config data or only read?

I'm loathe to break it, though. 😬